pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+        if amount_charged > Zero::zero() {
+            // We emit a payment charged event only if the user had to pay for its storage before the payment stream was updated
+            Self::deposit_event(Event::<T>::PaymentStreamCharged {
+                user_account: user_account.clone(),
+                backup_storage_provider_id: bsp_id,
+                amount: amount_charged,
+            });
+        }
+
+        // Check that the user is not flagged as without funds (after charging, to make sure it makes sense to update the payment stream)
+        ensure!(
+            !UsersWithoutFunds::<T>::contains_key(user_account),
+            Error::<T>::UserWithoutFunds
+        );
+
+        // Update the payment stream in the PaymentStreams mapping
+        PaymentStreams::<T>::mutate(bsp_id, user_account, |payment_stream| {
+            match payment_stream {
+                Some(payment_stream) => {
+                    payment_stream.rate = new_rate;
+                    Ok(())
+                }
+                None => {
+                    // This should never happen as we already checked that the payment stream exists
+                    return Err(Error::<T>::PaymentStreamNotFound);
+                }
+            }
+        })?;
+
+        Ok(())
+    }
+
+    /// This function holds the logic that checks if a payment stream can be removed and, if so, removes the payment stream from the PaymentStreams mapping,
+    /// decreases the user's payment streams count and, if the user has no more payment streams, releases the deposit.
+    pub fn do_delete_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+    ) -> DispatchResult {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Check that a payment stream between that BSP and user exists
+        ensure!(
+            PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamNotFound
+        );
+
+        // Charge the payment stream before deletion to make sure the storage provided by the provider is paid in full for its duration
+        let amount_charged = Self::do_charge_payment_stream(&bsp_account, user_account)?;
+        if amount_charged > Zero::zero() {
+            // We emit a payment charged event only if the user had to pay for its storage before deleting the payment stream
+            Self::deposit_event(Event::<T>::PaymentStreamCharged {
+                user_account: user_account.clone(),
+                backup_storage_provider_id: bsp_id,
+                amount: amount_charged,
+            });
+        }
+
+        // TODO: What do we do when a user is flagged as without funds? Does the provider assume the loss and we remove the payment stream?
```

Comment:
Yes, if the provider is executing this, then it means that he's taking the loss and it will stop storing the file for the user.

Reply:
Ok! I'll work on all the out-of-funds logic in another PR, this is already quite big as it is